### PR TITLE
SSO improvements main PR

### PR DESCRIPTION
**Pull request by alshakero**
This is the main PR towards SSO improvements. It's a place to merge PRs into, please do not push anything to it that is not updating it against trunk.

## Proposed changes:
Includes the following PRs:
- https://github.com/Automattic/jetpack/pull/35158
- https://github.com/Automattic/jetpack/pull/35282 
- https://github.com/Automattic/jetpack/pull/35298
- https://github.com/Automattic/jetpack/pull/35395
- https://github.com/Automattic/jetpack/pull/35234
- https://github.com/Automattic/jetpack/pull/35304
- https://github.com/Automattic/jetpack/pull/35473
- https://github.com/Automattic/jetpack/pull/35475
- https://github.com/Automattic/jetpack/pull/35531
- https://github.com/Automattic/jetpack/pull/35572
- https://github.com/Automattic/jetpack/pull/35573
- https://github.com/Automattic/jetpack/pull/35617
- https://github.com/Automattic/jetpack/pull/35683
- https://github.com/Automattic/jetpack/pull/35682
- https://github.com/Automattic/jetpack/pull/35685
- https://github.com/Automattic/jetpack/pull/35675

**Summary**:

 Site owners now have a much better experience managing users.
- Different colors for users connected, not invited or with a pending invite to join WordPress.com
- Possibility to quick send an invite to a user
- Possibility to revoke an invitation
- When adding a new user we now have a custom message sent, a notice that shows to the site owner that an invitation will be sent. And when the user is created, we send a WordPress.com invitation.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* Design/Kick-off thread: pdDR7T-1jU-p2
* Project proposal / feedback requested thread: pdDR7T-1ja-p2

## Testing instructions:

- Sync the Jetpack plugin to your site.
- Check that the summary changes are reflected.
- Try to break this.
- Check how it is on mobile.

### CODE REVIEW

**Comment by jeherve** [comment]
Is there a p2 post where I could find out more about this project? I'm curious what y'all have in mind for this feature :)

**Comment by agrullon95** [comment]
> Is there a p2 post where I could find out more about this project? I'm curious what y'all have in mind for this feature :)

@jeherve - You can read more about this project here: 
- Design/Kick-off thread: pdDR7T-1jU-p2
- Project proposal / feedback requested thread: pdDR7T-1ja-p2

**Comment by jeherve** [comment]
Thank you!

File: projects/plugins/jetpack/modules/sso.php [hunk 1]
```diff
@@ -81,6 +89,362 @@ private function __construct() {
...
+		return $actions;
+	}
+
+	/**
+	 * Render the invitation email message.
+	 */
+	public function render_invitation_email_message() {
+		$message = wp_kses(
+			__(
+				'New users will receive an invite to join WordPress.com, so they can log in securely using <a class="jetpack-sso-admin-create-user-invite-message-link-sso" rel="noopener noreferrer" target="_blank" href="https://jetpack.com/support/sso/">Secure Sign On</a>.',
```
Comment (heavyweight) [inline projects/plugins/jetpack/modules/sso.php]: Minor whine, but we could improve this to return support documentation based on the user's locale. For example https://jetpack.com/es/support/sso/ for spanish users

**Review by heavyweight — DISMISSED** [review]
I'm unable to edit my profile.
wp-admin/profile.php

Screenshot 2024-02-06 at 11 33 26

Screenshot 2024-02-06 at 11 35 27
Probably related



@agrullon95 : I'm working on this issue.

**Edit**
This is fixed now
✅

**Comment by heavyweight** [comment]
> I'm unable to edit my profile. wp-admin/profile.php
> 
> Screenshot 2024-02-06 at 11 33 26

Screenshot 2024-02-06 at 11 35 27
Probably related

**Comment by heavyweight** [comment]
https://github.com/Automattic/jetpack/assets/7000684/502a8544-cd62-4c20-ad01-123207c100b6

Created a new admin, visited the invitation link, and created my wp.com account. After that, I got redirected to the admin ui where I had that weird state as in the video. After logging in with wp.com again it got fixed.

To reproduce you will probably need to enable the classic admin setting in https://wpcalypso.wordpress.com/
Screenshot 2024-02-06 at 11 59 00

**Edit @alshakero**: this issue is documented 5158-gh-Automattic/dotcom-forge.

**Comment by heavyweight** [comment]
> Screen.Recording.2024-02-06.at.11.53.11.mov 
> Created a new admin, visited the invitation link, and created my wp.com account. After that, I got redirected to the admin ui where I had that weird state as in the video. After logging in with wp.com again it got fixed.

To reproduce you will probably need to enable the classic admin setting in https://wpcalypso.wordpress.com/
Screenshot 2024-02-06 at 11 59 00

**Comment by heavyweight** [comment]
Screenshot 2024-02-06 at 12 05 05
After creation the admin should manually send the invite? For the first admin user I created I think it was sent out automatically, but after that I only got the non-wpcom email and the functionality to invite the user by clicking the button

**Edit**
This is a non-issue please ignore it. The email were arriving late for me, and I got the email because I was admin which is expected
✅

**Comment by heavyweight** [comment]
> Screenshot 2024-02-06 at 12 05 05 After creation the admin should manually send the invite? For the first admin user I created I think it was sent out automatically, but after that I only got the non-wpcom email and the functionality to invite the user by clicking the button

Might be related
Screenshot 2024-02-06 at 12 21 51

**Edit**
Ignore this it is related to the issue above
✅

**Comment by heavyweight** [comment]
When we have a user with pending invite, if we delete him we should also "invalidate" the invite. Sure he could create a wp.com account but after he does that he should not be the admin on our site.
Imagine if I enter the email of my ex by mistake. I will panic for sure and delete the user. But after she creates her wpcom account she can come and take over my site 😈.

@alshakero: I'm working on this one.

**Comment by heavyweight** [comment]
I found where the notice came from.
In `send_wpcom_mail_user_invite` we try to find the user's locale:
`$locale  = get_user_locale( $user->ID )`
But for new users, this is the object shape: 
```
array(
   'user_login' => 'kokopapata8',
   'role' => 'administrator',
   'user_email' => 'k.papazov+8@gmail.com',
   'user_url' => '',
   'first_name' => 'koko',
   'last_name' => 'papata',
   'locale' => '',
   'comment_shortcuts' => '',
   'use_ssl' => 0,
   'user_pass' => 'H4X0R',
)
```

**Edit**
Addressed here
https://github.com/Automattic/jetpack/pull/35473/commits/416d7d05dad63b445f68410ec3be2ce070e0a8d0
✅

**Comment by heavyweight** [comment]
When creating a new user if there is an error, the values of the language selector and custom message will be lost.
You can reproduce this by submitting with an existing email or username and having all the form values filled.

@agrullon95 :  Fixed with https://github.com/Automattic/jetpack/pull/35531

**Comment by heavyweight** [comment]
https://github.com/Automattic/jetpack/assets/7000684/4464be44-d018-40dd-8d0e-de45cdc2ab87

The bottom two users were created before. I can invite, but can't revoke the invite. The first user is a user I created recently and it is working for him as expected

**Comment by heavyweight** [comment]
Tested visiting the invite link for a user that was deleted and thus with revoked permission.
Screenshot 2024-02-07 at 15 31 24
We can probably make some improvements here, but I don't see this as a blocker

File: projects/plugins/jetpack/modules/sso.php [hunk 2]
```diff
@@ -81,6 +91,442 @@ private function __construct() {
...
+			return self::create_error_notice_and_redirect( $query_params );
+		} else {
+			$query_params = array(
+				'jetpack-sso-invite-user'  => 'failed',
+				'jetpack-sso-invite-error' => 'invalid-user',
+				'_wpnonce'                 => $nonce,
+			);
+			return self::create_error_notice_and_redirect( $query_params );
+		}
+		wp_die();
```
Comment (heavyweight) [inline projects/plugins/jetpack/modules/sso.php]: We probably don't need this since it will be never reached

**Comment by escapemanuele** [comment]
> Tested visiting the invite link for a user that was deleted and thus with revoked permission.

Yes, I wouldn't make this blocking for this PR. Might be worth to address in the future as you said.

**Comment by heavyweight** [comment]
Noticed some inconsistencies with the locale settings.
What happens now is that the site language value is read and used when sending out the emails.
I changed the site language from `en` to `es` and created a user. The wp email was in `es` but the wpcom one was in `en`.
Screenshot 2024-02-07 at 22 26 56


Then I created a user with `en` language selected
Screenshot 2024-02-07 at 22 28 34
And got this
Screenshot 2024-02-07 at 22 29 22

Seem that our first wpcom email was possibly reading stale cache value.
It also seems the user locale setting is ignored for wp and wcom emails.

EDIT @heavyweight is working on this.

**Review by heavyweight — APPROVED** [review]
Re-tested and didn't find any issues

**Review by renancarvalho — APPROVED** [review]
Tested and did not locate any issues as well, I think we are good to go! 👍

**Pull request merged**